privacyidea/lib/tokens/foureyestoken.py

```diff
@@ -277,3 +330,182 @@ def authenticate(self, passw, user=None, options=None):
+        It then returns the the otp_counter = 1
+
+        :param user: the requesting user
+        :type user: User object
+        :param passw: the password: PIN + OTP
+        :type passw: string
+        :param options: additional arguments from the request, which could
+                        be token specific. Usually "transaction_id"
+        :type options: dict
+        :return: return 1 if the answer to the question is correct, -1 otherwise.
+        :rtype: int
+        """
+        options = options or {}
+        r_success = -1
+
+        # fetch the transaction_id
+        transaction_id = options.get('transaction_id')
+        if transaction_id is None:
+            transaction_id = options.get('state')
+
+        # get the challenges for this transaction ID
+        if transaction_id is not None:
+            challengeobject_list = get_challenges(serial=self.token.serial,
+                                                  transaction_id=transaction_id)
+
+            for challengeobject in challengeobject_list:
+                if challengeobject.is_valid():
+                    # challenge is still valid
+                    used_tokens = json.loads(challengeobject_list[0].data or json.dumps({}))
+                    remaining_realms = self._get_remaining_realms(used_tokens)
+                    r_success = self._authenticate_remaining_realms(passw, remaining_realms, used_tokens, options)
+
+                    if r_success:
+                        challengeobject.set_otp_status(True)
+                    if not r_success:
+                        # increase the received_count
+                        challengeobject.set_otp_status()
+
+        self.challenge_janitor()
+        return r_success
+
+    def create_challenge(self, transactionid=None, options=None):
+        """
+        This method creates a challenge, which is submitted to the user.
+        The submitted challenge will be preserved in the challenge
+        database.
+
+        The challenge is a randomly selected question of the available
+        questions for this token.
+
+        If no transaction id is given, the system will create a transaction
+        id and return it, so that the response can refer to this transaction.
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :type options: dict
+        :return: tuple of (bool, message, transactionid, attributes)
+        :rtype: tuple
+
+        The return tuple builds up like this:
+        ``bool`` if submit was successful;
+        ``message`` which is displayed in the JSON response;
+        additional ``attributes``, which are displayed in the JSON response.
+        """
+        options = options or {}
+        attributes = {}
+        message = ""
+        if type(options.get("data")) == dict:
+            # In the special first chal-resp case we do not have pickled data, yet. So we need to convert
```

Comment:
we don't pickle any more... Also we can still log a warning here if the jsonify results in string longer than 512 chars.

Reply:
No. I do not log!
I removed the wrapper function (which could have done the logging) 
So now I would have to add such logging at several places. This is why I refrain to log.

I change the "pickled" data in 57e8be50